lib/Sema/TypeCheckType.cpp

```diff
@@ -264,6 +264,11 @@ findDeclContextForType(TypeChecker &TC,
     if (ownerDC == parentDC)
       return std::make_tuple(parentDC, parentNominal, true);
 
+    if (isa<ExtensionDecl>(parentDC) && typeDecl == parentNominal) {
+      assert(parentDC->getParent()->isModuleScopeContext());
```

Comment:
An extension can appear inside another type, though. We diagnose it as invalid but still attempt to do name lookup into it.